File: doubly-linked-list/doubly-linked-list.py
class Node:
    def __init__(self, data, prev):
        self.data = data
        self.prev = prev
        self.next = None


class DoublyLinkedList:
    """
    A doubly-linked list data structure, where each node in the list contains references
    for both the node before it and the node after it.
    """

    def __init__(self):
        self.head = None
        self.tail = None

    def add(self, element):
        """Adds an element to the tail end of the list."""
        if self.tail:
            self.tail.next = Node(element, prev=self.tail)
            self.tail = self.tail.next
        else:
            self.head = Node(element, prev=None)
            self.tail = self.head

    def remove(self, element):
        """Removes all ocurrences of the given element from the list."""
        if self.head:
            # Remove as many heads as necessary
            while self.head.data == element:
                self.head = self.head.next
                self.head.prev = None

            # Remove as many tails as necessary
            while self.tail.data == element:
                self.tail = self.tail.prev
                self.tail.next = None

            # Iterate through rest of list
            current_node = self.head.next
            while current_node.next:
                if current_node.data == element:
                    # Link adjacent nodes
                    current_node.prev.next = current_node.next
                    current_node.next.prev = current_node.prev
                current_node = current_node.next

        else:
            return None

    def clear(self):
        """Clears the contents of the list."""
        self.head = self.tail = None

    def reverse(self):
        """Reverses the list in place (or returns `None` if the list is empty)."""
        if self.head:
            self.head = self.tail
            current_node = self.head
            while current_node.prev:
                # Swap adjacent nodes
                old_prev = current_node.prev
                current_node.prev = current_node.next
                current_node.next = old_prev

                current_node = current_node.next

            # Set new tail
            self.tail = current_node
            self.tail.prev = self.tail.next
            self.tail.next = None

        else:
            return None

File: doubly-linked-list/doubly-linked-list.ts
class Node {
  data: any;
  prev: Node;
  next: Node = null;

  constructor(data: any, prev: Node) {
    this.data = data;
    this.prev = prev;
  }
}

/**
 * A doubly-linked list data structure, where each node in the list contains
 * references for both the node before it and the node after it.
 */
export default class DoublyLinkedList {
  head: Node = null;
  tail: Node = null;

  /** Adds an element to the tail end of the list. */
  add(element: any): void {
    if (this.tail) {
      this.tail.next = new Node(element, this.tail);
      this.tail = this.tail.next;
    } else {
      this.head = new Node(element, null);
      this.tail = this.head;
    }
  }

  /**
   * Removes all occurrences of the given element from the list (or returns
   * `null` if the element is not present).
   */
  remove(element: any): void | null {
    if (this.head) {
      // Removes as many heads as necessary
      while (this.head.data === element) {
        this.head = this.head.next;
        this.head.prev = null;
      }

      // Removes as many tails as necessary
      while (this.tail.data === element) {
        this.tail = this.tail.prev;
        this.tail.next = null;
      }

      // Iterate through rest of list
      let currentNode = this.head.next;
      while (currentNode.next) {
        if (currentNode.data === element) {
          // Link adjacent nodes
          currentNode.prev.next = currentNode.next;
          currentNode.next.prev = currentNode.prev;
        }
        currentNode = currentNode.next;
      }
    } else {
      return null;
    }
  }

  /** Clears the contents of the list. */
  clear(): void {
    this.head = null;
    this.tail = null;
  }

  /** Reverses the list in place (or returns `null` if the list is empty). */
  reverse(): void | null {
    if (this.head) {
      this.head = this.tail;
      let currentNode = this.head;
      while (currentNode.prev) {
        // Swap adjacent nodes
        const oldPrev = currentNode.prev;
        currentNode.prev = currentNode.next;
        currentNode.next = oldPrev;

        currentNode = currentNode.next;
      }

      // Set new tail
      this.tail = currentNode;
      this.tail.prev = this.tail.next;
      this.tail.next = null;
    } else {
      return null;
    }
  }
}
